src/adaptors/adaptor_common.c

```diff
@@ -0,0 +1,270 @@
+            }
+        }
+
+
+        if (is_listener) {
+            if (config->authenticate_peer) {
+                res = pn_tls_config_set_peer_authentication(*tls_config, PN_TLS_VERIFY_PEER, config_ssl_profile->ssl_trusted_certificate_db);
+            }
+            else {
+                res = pn_tls_config_set_peer_authentication(*tls_config, PN_TLS_ANONYMOUS_PEER, 0);
+            }
+        }
+        else {
+            // Connector.
+            if (config->verify_host_name) {
+                res = pn_tls_config_set_peer_authentication(*tls_config, PN_TLS_VERIFY_PEER_NAME, config_ssl_profile->ssl_trusted_certificate_db);
+            }
+            else {
+                res = pn_tls_config_set_peer_authentication(*tls_config, PN_TLS_VERIFY_PEER, config_ssl_profile->ssl_trusted_certificate_db);
+            }
+
+            *tls_has_output = true; // always true for initial client side TLS.
+        }
+
+        if (res != 0) {
+            qd_log(log_source, QD_LOG_ERROR, "[C%"PRIu64"] Unable to set tls peer authentication for sslProfile %s - (%d)", conn_id, config->ssl_profile_name, res);
+            break;
+        }
+
+        //
+        // Provide an ordered list of application protocols for ALPN by calling pn_tls_config_set_alpn_protocols. In our case, h2 is the only supported protocol.
+        // A list of protocols can be found here - https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.txt
+        //
+        if (protocols)
+            pn_tls_config_set_alpn_protocols(*tls_config, protocols, NUM_ALPN_PROTOCOLS);
+
+        // set up tls session
+        if (*tls_session) {
+            pn_tls_free(*tls_session);
+        }
+        *tls_session = pn_tls(*tls_config);
+
+        int ret = pn_tls_start(*tls_session);
+        if (ret != 0) {
+            tls_setup_success = false;
+            break;
+        }
+
+        if (! *tls_session) {
+            qd_log(log_source, QD_LOG_ERROR, "[C%"PRIu64"] Unable to create tls session for sslProfile %s with hostname: '%s'", conn_id, config->ssl_profile_name, config->host);
+            break;
+        }
+        pn_tls_set_peer_hostname(*tls_session, config->host);
+
+        tls_setup_success = true;
+
+        qd_log(log_source, QD_LOG_INFO, "[C%"PRIu64"] Successfully configured ssl profile %s", conn_id, config->ssl_profile_name);
+
+
+    } while (0);
+
+    // Handle tls creation/setup failure by deleting any pn domain or session objects
+    if (!tls_setup_success) {
+        if (*tls_session) {
+            pn_tls_free(*tls_session);
+            *tls_session = 0;
+        }
+
+        if (*tls_config) {
+            pn_tls_config_free(*tls_config);
+            *tls_config = 0;
+        }
+
+    }
+
+    return tls_setup_success;
+}
+
+void qd_free_adaptor_config(qd_adaptor_config_t *config, qd_adaptor_type_t adaptor_type, qd_log_source_t  *log_source)
+{
+    if (!config)
+        return;
+
+    if (adaptor_type == TCP) {
```

Comment:
Why not put the adaptor_type into the qd_adaptor_config_t itself?  It would be useful to have it there, even if we only used it to assert() the correct type when accessing the proper union subfield.